### PR TITLE
Build script overhaul 2023

### LINKED ISSUE #28763: Release script can stomp over root `./deps.edn` dependencies and break the build
We ran into an issue running the release script today

```
Step failed: Syntax error compiling at (metabase/query_processor/middleware/cache/impl.clj:46:20).
{:via
 [{:type clojure.lang.Compiler$CompilerException,
   :message "Syntax error compiling at (metabase/query_processor/middleware/cache/impl.clj:46:20).",
   :data
   #:clojure.error{:phase :compile-syntax-check,
                   :line 46,
                   :column 20,
                   :source "metabase/query_processor/middleware/cache/impl.clj"},
   :at [clojure.lang.Compiler analyze "Compiler.java" 6825]}
  {:type java.lang.RuntimeException,
   :message "No such var: ordered-map/ordered-map-reader",
   :at [clojure.lang.Util runtimeException "Util.java" 221]}],
...
```

The root cause is that the version of `org.flatland/ordered` in the release script is stomping on the version in the root `deps.edn`

```
deps.edn
130:  org.flatland/ordered                      {:mvn/version "1.15.10"}            ; ordered maps & sets

bin/lint-migrations-file/deps.edn
5:  org.flatland/ordered {:mvn/version "1.5.9"}} ; used by io.forward/yaml -- need the newer version

bin/release/deps.edn
9:  org.flatland/ordered {:mvn/version "1.5.7"}

bin/build-mb/deps.edn
7:  org.flatland/ordered            {:mvn/version "1.15.10"}
```

We need to fix this issue and prevent it from happening again. We shouldn't be specifying these dependencies in multiple places

I think we can do one of the following:

1. Remove all the dependencies in the build/release scripts that are duplicated with ones in the root `deps.edn` file. But how do we prevent this from happening in the future? Seems to easy to accidentally put more deps in there in the future. I guess we'd have to write a linter -- ick

2. Do away with the separate `deps.edn` files for the build/release scripts, and just have them be different aliases inside the root `deps.edn`. Then something like `./bin/release.sh` would just do something like `clojure -X:build:release` using the `:build` and `:release` profiles from the root `deps.edn`

3. Building MB in a separate process is maybe doable but we have to build like 20 drivers in addition to metabase core itself and doing all that loopy logic in a separate process sounds wack. At that point the build code becomes a source file `:extra-paths` in an alias in the root `:deps.edn` which means we're basically doing the same thing as option (2) anyway
 
I'm leaning towards 2,  seems like the best option

### PR DESCRIPTION
**Pull request by camsaul**
Fixes #28763

### Esoteric details
I asked my tarot cards what card best represented this PR and I drew The Tower

and I was a little alarmed at first but I think it's actually a good omen that matches this PR perfectly. To quote [Tarot: Mirror of the Soul](https://www.amazon.com/Tarot-Mirror-Handbook-Aleister-Crowley/dp/0877286833)'s notes:

> Just as the extraction of a rotten tooth provides relief for the entire body, the destruction of stagnant situations .. begins a healing process for your entire organism. Having a tooth extracted can be painful, but when the tooth is poisoning your system, there is no other choice.

And that is exactly what we are doing here. An upheaval and transformation of the build scripts, destroying what's rotten and beginning anew

### Technical details

Previously the various build and release scripts were divided between 5 separate subprojects in `bin`, each with their own `deps.edn` files, **plus** `./build.clj`, which its own alias in `./deps.edn`, for a total of 6 places things lived. These different subprojects depended on one another and ultimately on the Metabase core project in the root of the repo as `:local/root` dependencies.

This mostly worked ok, but today we ran into a build script failure in #28763 because the build scripts were specifying an older version of `flatland.ordered` than the one in the root `deps.edn` file, and since we load up the whole system to build drivers, things failed as a result.

So the key goal here was to make sure none of the build/release subprojects had any dependencies that conflicted with stuff in the core project. Rather than come up with some hacky way to do that, it seemed like the appropriate fix was to just merge their `deps.edn` configurations into aliases inside `./deps.edn` itself. But then once I started doing that I realized it was ridiculous to have six different profiles with six different source paths that all depend on each other and share a bunch of code, so I just decided to collapse the `bin/.../src` and `bin/.../build` trees for the various build and release scripts into a combined `bin/build/src` and `bin/build/test` directories. We now have a few aliases like `:build` and `:build-dev` and `:build-test` in the root `deps.edn` file for working with this stuff and running tests, as well as a handful of ones like `:build/driver` for various entrypoints into that code.

So this PR really is mostly shuffling things around and merging stuff together, removing duplication so we don't have 4 different versions of the same dependency in different parts of the project. Ultimately it all works the same since we pulled all this stuff in when building in the first place.

I also moved `./build.clj` to `bin/build/src/build/uberjar.clj` so it could live with its friends. It hasn't been necessary to shell out to that namespace for a while now (there was a comment in the build script I added a while back to this effect), so we're calling it directly rather than spinning up a new process; this should speed up uberjar build time a little bit (maybe like 10-20 seconds or so, but it's tough to say). 

Besides that mostly everything all works the same, we were mostly using the shell scripts in `./bin` anyway, so I updated those to use the slightly different Clojure commands and everything is working.

Oh yeah, I renamed `./bin/build` to `./bin/build.sh` because I wanted to have `./bin/build` be the new place to stick the source code for this stuff. I updated references to it in GH actions so we should be g2g

### CODE REVIEW

File: .github/workflows/build-scripts.yml [hunk 1]
```diff
@@ -19,25 +19,8 @@ jobs:
...
-      timeout-minutes: 15
-    - name: Run build-mb build script tests
-      run: clojure -M:test
-      working-directory: bin/build-mb
-      timeout-minutes: 15
-    - name: Run release script tests
-      run: clojure -M:test
-      working-directory: bin/release
+    - name: Run build and release script tests
+      run: clojure -X:dev:drivers:build:build-dev:build-test
```
Comment (camsaul) [inline .github/workflows/build-scripts.yml:23]: I should also mention that we can run all the build/release script tests in one step instead of 5 separate steps now. Well we could have done this all along probably but it's much easier now. So this will be nice and speed up CI a bit. Not like this was super slow in the first place, but this will be a nice benefit

File: .github/workflows/e2e-tests.yml [hunk 2]
```diff
@@ -48,8 +48,8 @@ jobs:
         with:
           m2-cache-key: e2e-tests
 
-      - name: Build uberjar with ./bin/build
-        run: ./bin/build version translations frontend licenses drivers uberjar
```
Comment (camsaul) [inline .github/workflows/e2e-tests.yml:52]: There are only six steps you can run, and this lists all six. It's exactly the same as running without specifying any steps

On that note I'm not convinced we need to be doing all these steps and building the entire uberjar and all drivers and **checking licenses** in the 4 or 5 different e2e tests we do it in, but auditing that stuff is outside the scope of this PR

File: .github/workflows/percy-issue-comment.yml [hunk 3]
```diff
@@ -69,7 +69,7 @@ jobs:
 
       - name: Prepare cypress environment
         uses: ./.github/actions/prepare-cypress
-      - run: ./bin/build
+      - run: ./bin/build.sh
```
Comment (camsaul) [inline .github/workflows/percy-issue-comment.yml:72]: e.g. does percy issue comment need to be checking our library licenses, building every single driver, etc? Probably not.

File: bin/build-driver.sh [hunk 4]
```diff
@@ -19,5 +19,4 @@ check_clojure_cli
 source "./bin/clear-outdated-cpcaches.sh"
 clear_outdated_cpcaches
 
-cd bin/build-drivers
-clojure -M -m build-driver $@
+clojure -X:build:drivers:build/driver :driver $@
```
Comment (camsaul) [inline bin/build-driver.sh:22]: I switched some of these things to use cool new `-X` instead of gross old `-M`.

When I wrote all this `-M` stuff I was brand new to Clojure CLI. Now I actually know what I'm doing

File: bin/build-drivers/deps.edn [hunk 5]
```diff
@@ -1,40 +0,0 @@
...
-
- :aliases
- {:dev
-  {:extra-paths ["test"]}
-
-  :test
-  {:extra-paths ["test"]
-   :extra-deps  {com.cognitect/test-runner {:git/url "https://github.com/cognitect-labs/test-runner.git"
-                                            :sha     "209b64504cb3bd3b99ecfec7937b358a879f55c1"}}
-   :main-opts   ["-m" "cognitect.test-runner"]}}}
```
Comment (camsaul) [inline bin/build-drivers/deps.edn:40]: We don't need different test runners for the build scripts vs everything else now

File: bin/build/src/build.clj [hunk 6]
```diff
@@ -79,9 +80,7 @@
   {:pre [(#{:oss :ee} edition)]}
   (u/delete-file-if-exists! uberjar-filename)
   (u/step (format "Build uberjar with profile %s" edition)
-    ;; TODO -- we (probably) don't need to shell out in order to do this anymore, we should be able to do all this
-    ;; stuff directly in Clojure land by including this other `build` namespace directly (once we dedupe the names)
-    (u/sh {:dir u/project-root-directory} "clojure" "-T:build" "uberjar" :edition edition)
+    (uberjar/uberjar {:edition edition})
```
Comment (camsaul) [inline bin/build/src/build.clj:83]: Finally got to fix one of the million `TODO`s

File: .github/workflows/build-scripts.yml [hunk 7]
```diff
@@ -19,25 +19,8 @@ jobs:
...
-      timeout-minutes: 15
-    - name: Run build-mb build script tests
-      run: clojure -M:test
-      working-directory: bin/build-mb
-      timeout-minutes: 15
-    - name: Run release script tests
-      run: clojure -M:test
-      working-directory: bin/release
+    - name: Run build and release script tests
+      run: clojure -X:dev:drivers:build:build-dev:build-test:ci
```
Comment (camsaul) [inline .github/workflows/build-scripts.yml:23]: I should also mention that we can run all the build/release script tests in one step instead of 5 separate steps now. Well we could have done this all along probably but it's much easier now. So this will be nice and speed up CI a bit. Not like this was super slow in the first place, but this will be a nice benefit

File: bin/build-drivers.md [hunk 8]
```diff
@@ -36,10 +42,5 @@ clojure -M -m build-driver redshift
 Verify that a built driver looks correctly built.
 
 ```
-cd bin/verify-driver redshift
-clojure -M -m verify-driver redshift
-
-# or
-
-./bin/verify-driver redshift
+clojure -X:build:build/verify-driver :driver :mongo
```
Comment (camsaul) [inline bin/build-drivers.md:45]: `bin/verify-driver` hasn't existed since like 2020, so another nice benefit of this PR is that while I was updating other dox I updated this too.

File: deps.edn [hunk 9]
```diff
@@ -454,23 +457,121 @@
...
+  :build/release
+  {:exec-fn release/release}
+
+  ;; extra paths and deps for working on build scripts, or running the tests.
+  :build-dev
+  {:extra-paths ["bin/build/test"]
+
+   ;; some of these are here just for to have jars on the classpath for testing the licenses
+   ;; stuff (see [[build.licenses-test]])
+   :extra-deps
```
Comment (dpsutton) [inline deps.edn:556]: All of these were chosen as examples from our current deps. They aren't needed to be added.

File: resources/locales.clj [hunk 10]
```diff
@@ -23,6 +23,7 @@
    "es"
    "ja"
    "cs"
+   "ar"
```
Comment (dpsutton) [inline resources/locales.clj:26]: was this intentional? Seems weird for something new to show up here

Comment (camsaul) [reply resources/locales.clj:26]: oh yeah, not intentional, just a side-effect of me running scripts to test stuff. I'll revert

File: deps.edn (hunk 9 above)
Comment (camsaul) [reply deps.edn:554]: fixed

File: deps.edn [hunk 11]
```diff
@@ -231,7 +231,9 @@
                  ;; locally -- prefer debuggability over performance for local dev work.
                  "-XX:-OmitStackTraceInFastThrow"
                  ;; prevent Java icon from randomly popping up in macOS dock
-                 "-Djava.awt.headless=true"]}
+                 "-Djava.awt.headless=true"
+                 ;; include more details for debugging NPEs
+                 "-XX:+ShowCodeDetailsInExceptionMessages"]}
```
Comment (camsaul) [inline deps.edn:236]: Not really related, but this is a basic need for all users

File: .github/workflows/build-scripts.yml (hunk 7 above)
Comment (ariya) [reply .github/workflows/build-scripts.yml:23]: For practical purposes in the CI, it's easier to see the log (and search for errors etc) on GitHub Actions UI when there are X separate steps. Otherwise, everything is drawn in the log of one very big (and log) step.

Comment (camsaul) [reply .github/workflows/build-scripts.yml:23]: Sure, but it probably takes an extra 10 seconds to spin up Clojure in each step and load all the same namespaces up since they're all basically different entry points to the same project, so shaving a minute off of the CI job is worth it I think. Also people almost never touch this stuff so these tests almost never fail

**Comment by paoliniluis** [comment]
Added the change to the Dockerfile, many customers use that to build their own container image

File: deps.edn [hunk 12]
```diff
@@ -454,23 +459,119 @@
...
+   :exec-args {:only ["bin/build/test"]}}
+
 
 ;;; Other misc convenience aliases
 
   ;; Profile Metabase start time with clojure -M:profile
   :profile
   {:main-opts ["-m" "metabase.core" "profile"]
-   :jvm-opts  ["-XX:+CITime" ; print time spent in JIT compiler
+   :jvm-opts  ["-XX:+CITime"                                                    ; print time spent in JIT compiler
```
Comment (calherries) [inline deps.edn:574]:
```suggestion
   :jvm-opts  ["-XX:+CITime" ; print time spent in JIT compiler
```

**Review by calherries — APPROVED** [review]
I combed through this, including checking the deps were merged correctly. Looks good to go

**Comment by calherries** [comment]
There were just a couple of places  we still had `./bin/build`, which I updated.

File: deps.edn (hunk 12 above)
Comment (calherries) [reply deps.edn]: I committed this [whitespace](https://github.com/metabase/metabase/pull/28767/commits/39879a091076e81a56bfb4c5240200511d007a47)
